cmd/broker/ingress/main.go

```diff
@@ -150,9 +216,51 @@ func (f *Handler) dispatch(msg *provisioners.Message) error {
 type runnableServer struct {
 	logger *zap.Logger
 	s      *http.Server
+	// ShutdownTimeout is the duration to wait for the http.Server to gracefully
+	// shut down when the stop channel is closed. If this is zero or negative,
+	// shutdown will never time out.
+	// TODO alternative: zero shuts down immediately, negative means infinite
+	ShutdownTimeout time.Duration
+	// wg is a temporary workaround for Manager returning immediately without
+	// waiting for Runnables to stop. See
+	// https://github.com/kubernetes-sigs/controller-runtime/issues/350.
+	wg *sync.WaitGroup
 }
 
-func (r *runnableServer) Start(<-chan struct{}) error {
-	r.logger.Info("Ingress Listening...", zap.String("Address", r.s.Addr))
-	return r.s.ListenAndServe()
+func (r *runnableServer) Start(stopCh <-chan struct{}) error {
+	logger := r.logger.With(zap.String("address", r.s.Addr))
+	logger.Info("Listening...")
+
+	errCh := make(chan error)
+
+	go func() {
+		r.wg.Add(1)
+		err := r.s.ListenAndServe()
+		if err != http.ErrServerClosed {
+			errCh <- err
+		}
+	}()
+
+	var err error
+	select {
+	case err = <-errCh:
+		logger.Error("Error running HTTP server", zap.Error(err))
+	case <-stopCh:
+		var ctx context.Context
+		var cancel context.CancelFunc
+		if r.ShutdownTimeout > 0 {
+			ctx, cancel = context.WithTimeout(context.Background(), r.ShutdownTimeout)
+			defer cancel()
+		} else {
+			ctx = context.Background()
+		}
+		logger.Info("Shutting down...")
+		if err = r.s.Shutdown(ctx); err != nil {
+			logger.Error("Shutdown returned an error", zap.Error(err))
+		} else {
+			logger.Info("Shutdown done")
+		}
+	}
+	r.wg.Done()
```

Comment:
I _think_ this might panic if for some reason the stopCh was already closed. As the starting go routine might not have had time to start, so calling Done() may cause it to go below zero, and thus panic.

If you move the `r.wg.Add(1)` out of the goroutine, then I think it would work as expected. It would also make it easy to do:

```golang
r.wg.Add(1)
defer r.wg.Done()
```

Reply:
Good idea, done.